### ISSUE TITLE
Polish Leo's layout experience

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#4099]
**Update**: PR #4121 supersedes this issue.

From [this thread](https://groups.google.com/g/leo-editor/c/EiVIHM5tbjg/m/rkClIER2CQAJ):

- Create menu item that shows layout diagrams for all layouts.
- Make sure Leo uses a reasonable (legacy?) layout for newbies.
- Revise or delete info item #3982.

**Comment by tbpassin** [comment]
Here's my approach.  I think it's simple and reliable.  Please tell me what you think.  In the qt_layout.py module I create a new decorator: `@register_layout`. This stores the layout name and its docstring in a module-level dictionary.  We can decorate exactly the layout functions we want to appear in the listing of layout diagrams.  This way we don't have to write code the tries to guess which layout functions to include and which not to.

The layout cache class holds a reference to this dictionary.  So the command to show the listing of layouts can get them right from the cache instance, like this:

```
dw = c.frame.top
cache = dw.layout_cache
layouts = cache.layout_registry
listing = ''
for name, docstr in layouts.items():
    listing += f'{name}:\n' + '=' * len(name) + f'\n{docstr}\n'

g.es(listing, tabName='layouts')
```

Each layout function would use the decorator like this:
```
@g.command('layout-big-tree')
@register_layout('layout-big-tree')
def big_tree(event: LeoKeyEvent) -> None:
    """Create Leo's big-tree layout:
        ┌──────────────────┐
        │  outline         │
        ├──────────┬───────┤
        │  body    │  log  │
        ├──────────┴───────┤
        │  VR              │
        └──────────────────┘
    """
    # function code here
```

Anyone who wanted to add more layouts to the dictionary could get a reference to it from the cache instance, or by importing the qt_layout module and getting it from there.  It's declared this way:

`LAYOUT_REGISTRY = {}`

This code is already working.  I've only decorated a few layout commands so far.  Here's what the printout looks like:

```
layout-big-tree:
===============
Create Leo's big-tree layout:
        ┌──────────────────┐
        │  outline         │
        ├──────────┬───────┤
        │  body    │  log  │
        ├──────────┴───────┤
        │  VR              │
        └──────────────────┘
    
layout-fallback-layout:
======================
Apply a workable layout in case the layout setting is invalid.
layout-horizontal-thirds:
========================
Create Leo's horizontal-thirds layout:
        ┌───────────┬───────┐
        │  outline  │  log  │
        ├───────────┴───────┤
        │  body             │
        ├───────────────────┤
        │  VR               │
        └───────────────────┘
    
layout-legacy:
=============
Create Leo's legacy layout:
        ┌───────────┬──────┐
        │ outline   │ log  │
        ├───────────┼──────┤
        │ body      │ VR   │
        └───────────┴──────┘
```

**Comment by tbpassin** [comment]
The `register_layout` decorator can be used outside of qt_layout as well:

```
from leo.plugins.qt_layout import register_layout, LAYOUT_REGISTRY

@register_layout('trial registering')
def test_reg():
    """This is the docstring from test_reg."""
    g.es('This is test_reg()')

g.es(list(LAYOUT_REGISTRY.items())[-1])
# Prints ('trial registering', 'This is the docstring from test_reg.')
```

**Comment by tbpassin** [comment]
Currently we have two layout systems. The 6.8.1 system has commands that start with "use-" and my new one has commands that start with "layout-". I named mine this way 1) to distinguish them from the "use-" commands, and 2) I think it's more intuitive for a user to search for layout commands by typing "layout" rather than "use".  "use" is also more generic and more likely to be used by someone else for an unrelated purpose.

With the "use-" commands, the setting string should be the command name with "use-" removed, This needs to be explained to the user. The "use-" commands will reload the outline; the "layout-" commands will not.  This needs to be explained to the user.

Both sets of commands produce the same layouts if the prefix "use-" or "layout" is ignored.  This needs to be explained to the user.

The new layout system has a command "layout-restore-default".  This commands looks at the setting  `qt-layout-name`.  If the setting's value does not start with "layout", then add the prefix, see if there is a command by that name.  If so, run that command.  Otherwise see if there is an unprefixed command with the setting's value.  If so, run that.

In summary of the last paragraph: the new layout command "layout-restore-default" can use the bare name of a layout ("vertical-thirds") with or without a "layout-" prefix and it will reproduce the same layout as the matching "use-" command produced.  This makes the new system completely compatible with the old.

**Comment by edreamleo** [comment]
On Tue, Oct 15, 2024 at 7:55 PM Thomas Passin ***@***.***>
wrote:

> Currently we have two layout systems.
>
I want a single system, even if it breaks existing code.

However, commands can have aliases.

@g.command('name-1')
@g.command('name-2')
def doMyCommand...
   ...

Otoh, I see no reason to retain the 'use-*' commands.

Let's keep layouts as simple as possible!

Edward

**Comment by tbpassin** [comment]
> I want a single system, even if it breaks existing code.

Same here. I didn't know if you wanted to make the change-over in two stages or one.

Here's how I think the new system could be used at outline load time.  Create the main and secondary splitters as they are now but make them children of the layout cache widget instead of where-ever they go now.  Then execute the 'layout-restore-default' command. This sounds dead-simple to me.

I think that initialization is far enough along at this point that it will work (but I haven't tested it yet). If not, wait until it has progressed further before executing the command.

I'm in the middle of getting my branch ready for a PR.  The branch contains the new decorator and uses it to build the dictionary of commands that we will use to create the help-for pane.  I'm having a devil of a time with mypy and it's delaying the PR.  I'd like to get this PR done before making one for the startup changes.  Is that OK with you?  It should only be a few more days.

**Comment by edreamleo** [comment]
On Wed, Oct 16, 2024 at 9:14 AM Thomas Passin ***@***.***>
wrote:

> I want a single system, even if it breaks existing code.
>
> Same here. I didn't know if you wanted to make the change-over in two
> stages or one.
>
Do whatever is easiest for you, preferring a single stage if possible.

> Here's how I think the new system could be used at outline load time.
> Create the main and secondary splitters as they are now but make them
> children of the layout cache widget instead of where-ever they go now. Then
> execute the 'layout-restore-default' command. This sounds dead-simple to me.
>
I don't know about the details to comment, but dead-simple always works for
me :-)

> I think that initialization is far enough along at this point that it will
> work (but I haven't tested it yet). If not, wait until it has progressed
> further before executing the command.
>
> I'm in the middle of getting my branch ready for a PR. The branch contains
> the new decorator and uses it to build the dictionary of commands that we
> will use to create the help-for pane.
>
Sounds good.

> I'm having a devil of a time with mypy and it's delaying the PR.
>
Alright. I can fix mypy problems in my separate R.

> I'd like to get this PR done before making one for the startup changes. Is
> that OK with you? It should only be a few more days.
>
Yes, it's OK with me. We get no brownie points for doing everything in one
PR.

Edward

**Comment by tbpassin** [comment]
Partly fulfilled by PR https://github.com/leo-editor/leo-editor/pull/4103.

Wait until you see how I solved mypy complaints! These techniques should be applicable to other parts of Leo's code that are now `typing: ignore`d.

**Comment by edreamleo** [comment]
@tbpassin Using TypeVar is elegant. Good work.

**Comment by tbpassin** [comment]
Here is one way to generate a help-for message in the Log frame for the built-in layouts:

```
dw = c.frame.top
cache = dw.layout_cache
layouts = cache.layout_registry
listing = ''
for name, docstr in layouts.items():
    listing += f'{name}:\n' + '=' * len(name) + f'\n{docstr}\n'

g.es(listing, tabName='layouts')
```

> Here's how I think the new system could be used at outline load time. Create the main and secondary splitters as they are now but make them children of the layout cache widget instead of where-ever they go now. Then execute the 'layout-restore-default' command. This

I have this basically working.  There are a few quirps to work through, as always.

**Comment by edreamleo** [comment]
@tbpassin Keep up the good work!

**Comment by edreamleo** [comment]
Closed in favor of PR #4121.
